Creating a specs folder, adding first spec (CommandBar features) (#4413)
microsoft/microsoft-ui-xaml: specs/CommandBar/new-command-bar-features.md

                <SplitButton.Flyout>
                    <MenuFlyout Placement="RightEdgeAlignedTop">
                        <MenuFlyoutItem Text="Insert above"/>
                        <MenuFlyoutItem Text="Insert between"/>
                        <MenuFlyoutItem  Text="Insert below"/>
                    </MenuFlyout>
                </SplitButton.Flyout>
            </SplitButton>
        </AppBarElementContainer>
        <AppBarButton Label="Select all"/>
        <AppBarButton Label="Delete" Icon="Delete"/>
    </CommandBarFlyout.SecondaryCommands>
</CommandBarFlyout>
```

### Feature #2. `RadioMenuFlyoutSubItem`

`RadioMenuFlyoutSubItem` is a new class that represents a "parent" `RadioMenuFlyout` item. 
It is very similar to the [MenuFlyoutSubItem](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.menuflyoutsubitem?view=winrt-19041) class.

All child items of a `RadioMenuFlyoutSubItem` should be of type 
[RadioMenuFlyoutItem](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.radiomenuflyoutitem?view=winui-2.5) 
(although this is not enforced by the API). If a child item is selected, the parent item will show as selected. 

> All child items should have the same GroupName as siblings of the parents. 

Reviewer: Discussion: What if you have two flyouts, all with the same group name? 
It's everything with the same GroupName within the whole menu.

Recommend: explain this

A `CommandBar` provides access to app-level or page-specific commands. 
A `CommandBarFlyout` is a `CommandBar` that appears in a `Flyout` menu, and must be invoked.

*CommandBar:*

![CommandBar example](images/commandbar-example.png)

*CommandBarFlyout* (invoked for example with a mouse right click):

![CommandBarFlyout example](images/commandbarflyout-example.png)

### Issue #1: You can only use a select few object types inside of a CommandBar/CommandBarFlyout. 

When using the `CommandBar`or `CommandBarFlyout` controls today, the only supported types for items are 
[AppBarButton](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.appbarbutton?view=winrt-19041), 
[AppBarToggleButton](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.appbartogglebutton?view=winrt-19041), 
and [AppBarSeparator](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.appbarseparator?view=winrt-19041). 
When you use an AppBarButton in a `CommandBar` or `CommandBarFlyout`, all of the interactions and styles needed to conform with

> the parent control are included - height, width, the background color when hovered/pressed, the font size, etc. 

Reviewer 2: Took me a while to figure out what "included" meant.

```suggestion
the parent control are provided automatically - height, width, the background color when hovered/pressed, the font size, etc. 
```

Reviewer: Recommend

    </MenuFlyoutSubItem>
</MenuFlyout>
```

The `MenuFlyoutSubItem` class represents a "parent" MenuFlyout item. It has an
[Items](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.menuflyout.items?view=winrt-19041) 
property, where you can assign all of its child items.  Here's the above example with labels showing which item is which type:

![Anatomy of a MenuFlyout](images/anatomy-diagram.PNG)

You also have the option of using 
[RadioMenuFlyoutItem](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.radiomenuflyoutitem?view=winui-2.5), 
which has a [GroupName](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.radiomenuflyoutitem.groupname?view=winui-2.5#Microsoft_UI_Xaml_Controls_RadioMenuFlyoutItem_GroupName) 
property. Items with the same GroupName will be a part of the same selection model - only one of the items in a Group can be "selected" and display the selection indicator at a time:

```xaml
<AppBarButton>View
    <AppBarButton.Flyout>
        <MenuFlyout>
            <MenuFlyoutItem>Open</MenuFlyoutItem>
            <MenuFlyoutSeparator/>

>             <RadioMenuFlyoutItem>Landscape</RadioMenuFlyoutItem>
>             <RadioMenuFlyoutItem IsChecked="True">Portrait</RadioMenuFlyoutItem>
>             <MenuFlyoutSeparator/>
>             <RadioMenuFlyoutItem>Small icons</RadioMenuFlyoutItem>
>             <RadioMenuFlyoutItem IsChecked="True">Medium icons</RadioMenuFlyoutItem>
>             <RadioMenuFlyoutItem>Large icons</RadioMenuFlyoutItem>

Reviewer 2: Missing `GroupName` attributes on these items. Kind of important since that's the purpose of the example.

```suggestion
            <RadioMenuFlyoutItem GroupName="Orientation">Landscape</RadioMenuFlyoutItem>
            <RadioMenuFlyoutItem GroupName="Orientation" IsChecked="True">Portrait</RadioMenuFlyoutItem>
            <MenuFlyoutSeparator/>
            <RadioMenuFlyoutItem GroupName="IconSize">Small icons</RadioMenuFlyoutItem>
            <RadioMenuFlyoutItem GroupName="IconSize" IsChecked="True">Medium icons</RadioMenuFlyoutItem>
            <RadioMenuFlyoutItem GroupName="IconSize">Large icons</RadioMenuFlyoutItem>
```

expand the list of secondary commands). The developer is currently not able to control how the `CommandBarFlyout` appears when invoked, 
and cannot control whether it can be collapsed or not. 

# Description

### Feature #1. Built-in styles to support adding `SplitButton` to `CommandBar` and `CommandBarFlyout`

WinUI will provide built in styles for SplitButton that will allow them to conform with the look and feel of an `AppBarButton`. 
These styles will change properties such as the `SplitButton`'s height/width, font size, color, interactions (color changes), and more. 

There will be two styles provided: `CommandBarSplitButtonStyle` and `CommandBarFlyoutSplitButtonStyle`. 
These styles can be accessed as ThemeResources in any WinUI app. 

Note: if a SplitButton item inside of a CommandBar enters the CommandBar's overflow menu, it will pick up the `CommandBarFlyoutSplitButtonStyle`.

For example:

```xml
<CommandBarFlyout x:Name="ImageCommandsFlyout" >  
    <AppBarButton Icon="Copy" ToolTipService.ToolTip="Copy" Label="Copy"/>

    <CommandBarFlyout.SecondaryCommands>
        <AppBarElementContainer>
            <SplitButton ToolTipService.ToolTip="Insert"

>                          Style="{ThemeResource SplitButtonCommandBarFlyoutStyle}">

Reviewer 2: Is there some way to get the styling automatically, or does everybody have to remember to opt into the "correct" style manually? Would it be overkill to have a `AppBarSplitButton` element that applies the style for you?

```csharp
unsealed runtimeclass AppBarSplitButton : SplitButton
{
  // no new members. Default Style is SplitButtonCommandBarFlyoutStyle.
}
```

Reviewer 3: We could have a setter inside the CommandBarFlyout resources for the SplitButton (and other controls). Those styles would only apply to the controls inside the CommandBarFlyout (I hope, flyouts are always a bit tricky in that area). That way we wouldn't need wrappers of have people remember to opt into the correct style.

Author: Good callout - when you place a SplitButton in the CommandBar or CommandBarFlyout, the correct style will automatically be applied.

Reviewer: Discussion: Yes, when you put these controls in a CommandBar, it automatically picks up the style. So the only thing special you need to know to use these controls is to put them into the AppBarElementContainer.

Recommend: Update to explain that, and take it out of the example code.

                    </StackPanel>
                </SplitButton.Content>
                <SplitButton.Flyout>
                    <MenuFlyout Placement="RightEdgeAlignedTop">
                        <MenuFlyoutItem Text="Insert above"/>
                        <MenuFlyoutItem Text="Insert between"/>
                        <MenuFlyoutItem  Text="Insert below"/>
                    </MenuFlyout>
                </SplitButton.Flyout>
            </SplitButton>
        </AppBarElementContainer>
        <AppBarButton Label="Select all"/>
        <AppBarButton Label="Delete" Icon="Delete"/>
    </CommandBarFlyout.SecondaryCommands>
</CommandBarFlyout>
```

### Feature #2. `RadioMenuFlyoutSubItem`

`RadioMenuFlyoutSubItem` is a new class that represents a "parent" `RadioMenuFlyout` item. 
It is very similar to the [MenuFlyoutSubItem](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.menuflyoutsubitem?view=winrt-19041) class.

All child items of a `RadioMenuFlyoutSubItem` should be of type 
[RadioMenuFlyoutItem](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.radiomenuflyoutitem?view=winui-2.5) 

> (although this is not enforced by the API). If a child item is selected, the parent item will show as selected. 

Reviewer 2: Can I at least put a separator in there?

Some people like to put dummy elements inside a flyout, e.g., to provide labels. Is that allowed?
```xml
<RadioMenuFlyoutSubItem>
    <MenuFlyoutItem IsEnabled="False">Other free options</MenuFlyoutItem>
    <RadioMenuFlyoutItem GroupName="Shipping">Enhanced shipping</RadioMenuFlyoutItem>
    <MenuFlyoutSeparator/>
    <MenuFlyoutItem IsEnabled="False">Extra charge</MenuFlyoutItem>
    <RadioMenuFlyoutItem GroupName="Shipping">Premium shipping</RadioMenuFlyoutItem>
    <RadioMenuFlyoutItem GroupName="Shipping">Overnight shipping</RadioMenuFlyoutItem>
</RadioMenuFlyoutSubItem>
```

Reviewer: Discussion: these two examples should work (separator and the label).

Recommend: make the recommendation more nuanced, or just a suggestion, etc.

> All child items should have the same GroupName as siblings of the parents. 

Reviewer 2: So this isn't allowed?

```xml
<MenuFlyout>
        <RadioMenuFlyoutItem GroupName="Orientation">Landscape</RadioMenuFlyoutItem>
        <RadioMenuFlyoutItem GroupName="Orientation" IsChecked="True">Portrait</RadioMenuFlyoutItem>
        <MenuFlyoutSeparator/>
        <RadioMenuFlyoutItem GroupName="IconSize" IsChecked="True">Small icons</RadioMenuFlyoutItem>
        <RadioMenuFlyoutSubItem Text="Other sizes">
            <RadioMenuFlyoutItem GroupName="IconSize">Medium icons</RadioMenuFlyoutItem>
            <RadioMenuFlyoutItem GroupName="IconSize">Large icons</RadioMenuFlyoutItem>
        </RadioMenuFlyoutSubItem>
</MenuFlyout>
```

Reviewer: Discussion: This is allowed.

Recommend: update the description (have the same name as at least one of the parent's siblings).

                <RadioMenuFlyoutItem Text="Name" GroupName="SortOption"/>
                <RadioMenuFlyoutItem Text="Date" GroupName="SortOption" 
                                     IsChecked="True"/>
                <RadioMenuFlyoutItem Text="Size" GroupName="SortOption"/>
                <RadioMenuFlyoutSubItem Text="Other">
                    <RadioMenuFlyoutItem Text="Album name" GroupName="SortOption"/>
                    <RadioMenuFlyoutItem Text="Artist name" GroupName="SortOption"/>
                    <RadioMenuFlyoutItem Text="Genre" GroupName="SortOption"/>
                </RadioMenuFlyoutSubItem>
            </MenuFlyout>
        </AppBarButton.Flyout>
    </AppBarButton>

    <AppBarSeparator/>
    <AppBarButton Icon="Play" Label="Play" />
    <AppBarButton Icon="Forward" Label="Forward" />

</CommandBar>
```

### Feature #3. AlwaysExpanded API

This will be a new API added to `CommandBarFlyout` that will give the developer the ability to keep the `CommandBarFlyout` in its expanded mode at all times.

When the AlwaysExpanded property is set to true, the [...] button will not appear, 
and the user will not be able to collapse the `CommandBarFlyout`. 

> Once the [...] button is collapsed, other `AppBarButtons`/`CommandBarFlyout` items should be able to take its space. 

Reviewer 2: Potentially confusing reuse of "collapsed".

```suggestion
If the [...] button is not visible, then other `AppBarButtons`/`CommandBarFlyout` items are able to take its space. 
```

Reviewer: Recommend

This property will only have an effect if the `CommandBarFlyout` has secondary commands. 
If there are no secondary commands, the `CommandBarFlyout` will always be in collapsed mode (hence why there’s no `AlwaysCollapsed` property). 

> `CommandBarFlyout` can still be collapsed/expanded by the developer programmatically even when this property is set to true. When an item in the primary commands list is invoked, the CommandBarFlyout should stay open and expanded. If an item in the secondary commands list is invoked, the CommandBarFlyout should close completely. 

Reviewer 2: Is this really two paragraphs?

1. Programmatic collapse/expand.
2. Autocollapse behavior.

Reviewer: Recommend: yes, please, two paragraphs

> `CommandBarFlyout` can still be collapsed/expanded by the developer programmatically even when this property is set to true. When an item in the primary commands list is invoked, the CommandBarFlyout should stay open and expanded. If an item in the secondary commands list is invoked, the CommandBarFlyout should close completely. 

Reviewer 2: > ... the CommandBarFlyout should stay open and expanded.

I think this should be "should stay open and not change its collapsed/expanded state." Because picking a primary command from a collapsed CommandBarFlyout presumably does not expand it.

Reviewer: Recommend

For example:

```xaml
<CommandBarFlyout x:Key="Flyout1" AlwaysExpanded='true'>
    <AppBarButton Label="Share" Icon="Share" Click="OnElementClicked" />
    <AppBarButton Label="Save" Icon="Save" Click="OnElementClicked" />
    <CommandBarFlyout.SecondaryCommands>
        <AppBarButton x:Name="ResizeButton1" Label="Resize" Click="OnElementClicked" />
        <AppBarButton x:Name="MoveButton1" Label="Move" Click="OnElementClicked" />
    </CommandBarFlyout.SecondaryCommands>
</CommandBarFlyout>

> ```

Reviewer 2: Probably need a picture here. Because the markup by itself doesn't demonstrate the behavior.

Reviewer: Recommend: please

| RadioMenuFlyoutSubItemForegroundDisabled    | Text color when disabled                          |
| RadioMenuFlyoutSubItemChevron               | Chevron color at rest                             |
| RadioMenuFlyoutSubItemChevronPointerOver    | Chevron color on hover                            |
| RadioMenuFlyoutSubItemChevronPressed        | Chevron color when pressed                        |
| RadioMenuFlyoutSubItemChevronDisabled       | Chevron color when disabled                       |
| RadioMenuFlyoutSubItemChevronSubMenuOpened  | Chevron color when opened                         |

# API Details

```csharp
unsealed runtimeclass CommandBarFlyout : Windows.UI.Xaml.Controls.Primitives.FlyoutBase
{
    // Existing members elided 
    // ...

    boolean AlwaysExpanded { get; set; }
    
    static Windows.UI.Xaml.DependencyProperty AlwaysExpandedProperty { get; }
};
```

```csharp
unsealed runtimeclass RadioMenuFlyoutSubItem : Microsoft.UI.Xaml.Controls.MenuFlyoutItemBase
{
    RadioMenuFlyoutSubItem();

    IList<MenuFlyoutItemBase> Items { get; }
    IconElement Icon { get; set; }
    string Text { get; set; }

>     static Windows.UI.Xaml.DependencyProperty Items { get; }
>     static Windows.UI.Xaml.DependencyProperty Icon { get; }
>     static Windows.UI.Xaml.DependencyProperty Text { get; }

Reviewer 2: ```suggestion
    static Windows.UI.Xaml.DependencyProperty ItemsProperty { get; }
    static Windows.UI.Xaml.DependencyProperty IconProperty { get; }
    static Windows.UI.Xaml.DependencyProperty TextProperty { get; }
```

Reviewer: Recommend
